From ba075274a7acc99bb15e3c3306b2768356aa0459 Mon Sep 17 00:00:00 2001
From: Lin Yuan <linyuan@google.com>
Date: Thu, 26 May 2022 18:49:21 -0400
Subject: [PATCH] RESTRICT AUTOMERGE Fix: policy enforcement for location wifi
 scanning

Make DISALLOW_CONFIG_LOCATION effectively disallow wifi scanning and
bluetooth scanning settings for location services.

screenshots: http://shortn/_EUOdqrOcnS, http://shortn/_j320QDm1Zo

Bug: 228315522
Bug: 228315529
Test: atest SettingsRoboTests, on device

Change-Id: I78291579a79e915a27ebdd051b3caf3fc04efc41
(cherry picked from commit fcae147f58be3b6441ce9e03bc59515af0d53ccc)
(cherry picked from commit 2598a722aef1a65fec2d9b66baf9e800fc72a7b5)
Merged-In: I78291579a79e915a27ebdd051b3caf3fc04efc41
---
 res/xml/location_services.xml                         |  4 ++--
 ...ServicesBluetoothScanningPreferenceController.java | 11 +++++++++++
 ...ationServicesWifiScanningPreferenceController.java | 11 +++++++++++
 3 files changed, 24 insertions(+), 2 deletions(-)

diff --git a/res/xml/location_services.xml b/res/xml/location_services.xml
index 2de4e7b244..04eff2343a 100644
--- a/res/xml/location_services.xml
+++ b/res/xml/location_services.xml
@@ -23,13 +23,13 @@
             android:layout="@layout/preference_category_no_label"
             settings:controller="com.android.settings.location.LocationInjectedServicesPreferenceController"/>
 
-        <Preference
+        <com.android.settingslib.RestrictedPreference
             android:fragment="com.android.settings.location.WifiScanningFragment"
             android:key="location_services_wifi_scanning"
             android:title="@string/location_scanning_wifi_always_scanning_title"
             settings:controller="com.android.settings.location.LocationServicesWifiScanningPreferenceController"/>
 
-        <Preference
+        <com.android.settingslib.RestrictedPreference
             android:fragment="com.android.settings.location.BluetoothScanningFragment"
             android:key="location_services_bluetooth_scanning"
             android:title="@string/location_scanning_bluetooth_always_scanning_title"
diff --git a/src/com/android/settings/location/LocationServicesBluetoothScanningPreferenceController.java b/src/com/android/settings/location/LocationServicesBluetoothScanningPreferenceController.java
index 9913848eb2..ccda5722b0 100644
--- a/src/com/android/settings/location/LocationServicesBluetoothScanningPreferenceController.java
+++ b/src/com/android/settings/location/LocationServicesBluetoothScanningPreferenceController.java
@@ -16,10 +16,14 @@
 package com.android.settings.location;
 
 import android.content.Context;
+import android.os.UserManager;
 import android.provider.Settings;
 
+import androidx.preference.Preference;
+
 import com.android.settings.R;
 import com.android.settings.core.BasePreferenceController;
+import com.android.settingslib.RestrictedPreference;
 
 /**
  * Preference controller for Bluetooth scanning in Location Services.
@@ -31,6 +35,13 @@ public class LocationServicesBluetoothScanningPreferenceController extends
         super(context, key);
     }
 
+    @Override
+    public void updateState(Preference preference) {
+        ((RestrictedPreference) preference).checkRestrictionAndSetDisabled(
+                UserManager.DISALLOW_CONFIG_LOCATION);
+        refreshSummary(preference);
+    }
+
     @Override
     public CharSequence getSummary() {
         final boolean bleScanOn = Settings.Global.getInt(mContext.getContentResolver(),
diff --git a/src/com/android/settings/location/LocationServicesWifiScanningPreferenceController.java b/src/com/android/settings/location/LocationServicesWifiScanningPreferenceController.java
index 080a023268..2814f9d7f0 100644
--- a/src/com/android/settings/location/LocationServicesWifiScanningPreferenceController.java
+++ b/src/com/android/settings/location/LocationServicesWifiScanningPreferenceController.java
@@ -17,9 +17,13 @@ package com.android.settings.location;
 
 import android.content.Context;
 import android.net.wifi.WifiManager;
+import android.os.UserManager;
+
+import androidx.preference.Preference;
 
 import com.android.settings.R;
 import com.android.settings.core.BasePreferenceController;
+import com.android.settingslib.RestrictedPreference;
 
 /**
  * Preference controller for Wi-Fi scanning in Location Services.
@@ -33,6 +37,13 @@ public class LocationServicesWifiScanningPreferenceController extends BasePrefer
         mWifiManager = context.getSystemService(WifiManager.class);
     }
 
+    @Override
+    public void updateState(Preference preference) {
+        ((RestrictedPreference) preference).checkRestrictionAndSetDisabled(
+                UserManager.DISALLOW_CONFIG_LOCATION);
+        refreshSummary(preference);
+    }
+
     @Override
     public CharSequence getSummary() {
         final boolean wifiScanOn = mWifiManager.isScanAlwaysAvailable();
-- 
2.37.0.rc0.161.g10f37bed90-goog

